
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Perturb-seq data analysis &#8212; RegVelo&#39;s reproducibility repository</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'zebrafish/8_perturb_seq_analysis';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="elf1 perturbation simulation" href="9_elf1_motif_analysis.html" />
    <link rel="prev" title="Perturbation prediction benchmark" href="7_perturbation_benchmark.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">RegVelo's reproducibility repository</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    RegVelo’s reproducibility repository
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Complexity scale analysis with mouse neural crest and Schwann cell datasets</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/1_split_scaling.html">Split mouse neural crest data into different scale</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/2_scenic_stage.html">Construct pySCENIC GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/3_grn_preprocessing.html">Preprocess GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/3_regvelo.html">RegVelo-based analysis on murine NCC cell</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/3_scvelo.html">scVelo application on murine neural crest data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/3_velovi.html">veloVI application on murine neural crest data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/4_cbc.html">Benchmark transition from hub cells to terminal states</a></li>

<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/4_cell_fate_consistency.html">Evaluate consistency of velocity and cell fate probability across scale</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scale_mouse_neural_crest/4_tsi.html">Benchmark Terminal State Identification</a></li>

</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Cell cycle (U2OS)</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/1_data_preparation.html">Cell cycle data preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_regvelo.html">RegVelo benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_scvelo.html">scVelo benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_velovi.html">veloVI benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_tfvelo.html">TFvelo benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_unitvelo.html">UniTVelo benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_velovae_fullvb.html">veloVAE (fullvb) benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_velovae.html">veloVAE (VAE) benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_cell2fate.html">cell2fate benchmark on cell cycle data</a></li>






<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_dpt.html">DPT benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_celloracle.html">CellOracle benchmark on cell cycle</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_grnboost2.html">GRNBoost2 benchmark on cell cycle</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_splicejac.html">SpliceJAC benchmark on cell cycle</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/2_correlation.html">Correlation benchmark on cell cycle</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/3_comparison.html">Performance comparison of inference on cell cycle</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/3_grn_influence.html">GRN influence on velocity inference with RegVelo</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle/3_grn_visualization.html">Visualize GRN</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Cell cycle (RPE1)</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle_rpe1/1_data_preparation_rpe1.html">Cell cycle data preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle_rpe1/2_regvelo.html">RegVelo benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle_rpe1/3_celldancer.html">CellDancer benchmark on cell cycle data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle_rpe1/4_alpha_prediction.html">Benchmark transcription rate prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cell_cycle_rpe1/5_latent_time.html">Evaluate latent time prediction via redefining global latent time</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Pancreatic endocrinogenesis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/1_preprocessing.html">Basic preprocessing of the pancreatic endocrine dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/2_regvelo.html">RegVelo-based analyis of pancreatic endocrine data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/2_scvelo.html">scVelo-based analyis of pancreatic endocrine data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/3_comparison_tsi.html">TSI comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/4_cell_fate_analysis.html">Pancreatic endocrine cell fate analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/4_test_gene_regulation.html">Regulation is important for terminal state identification in pancreatic endocrine</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/5_cell_cycle_analysis.html">E2f1 perturbation on pancreatic endocrine dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/6_e2f1_targets_analysis.html">Evaluate RegVelo predicted E2f1 downstream targets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/7_repeat_regvelo.html">Robustness of RegVelo’s perturbation analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/8_perturbation_screening.html">Perturbation prediction on pancreatic endocrine via RegVelo</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/9_repeat_velovi.html">veloVI-based analyis of pancreatic endocrine data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/10_benchmark_identifiability.html">Benchmark cell fate and velocity identifiability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/11_neurod2_analysis.html">Analyzing perturbation simulation results on pancreatic endocrine dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pancreatic_endocrinogenesis/11_screening_neurod2_targets.html">Neurod2 target analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Hematopoiesis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/1_preprocessing.html">Preprocessing of the human hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/2_regvelo.html">Regularized RegVelo-based analysis of hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/2_regvelo_unregularized.html">RegVelo-based analysis of hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/2_scvelo.html">scVelo-based analysis of hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/2_velovi.html">veloVI-based analysis of hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/3_comparison_gene_ranking.html">Comparing lineage-associated gene rankings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/3_comparison_uncertainty.html">Comparing uncertainty</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/3_comparison_velocity.html">Performance comparison of inference on hematopoiesis dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/4_dynamo.html">Dynamo-based drivers analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/4_perturbation.html">Perturbation prediction on hematopoiesis dataset via RegVelo</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/5_comparison_drivers.html">Comparison of driver TF ranking performance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/6_cell_fate_motif.html">PU.1 and GATA1 perturbation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/7_robustness.html">Repeat run RegVelo and veloVI</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hematopoiesis/8_comparison_robustness.html">Compare estimated velocity robustness between RegVelo and veloVI</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Zebrafish neural crest</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="1_ss3_analysis.html">Basic preprocessing and analysis of the zebrafish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_regvelo.html">RegVelo-based analyis of zebrafish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_scvelo.html">scVelo-based analyis of zebrafish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_velovi.html">veloVI-based analyis of zebrafish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_tsi_comparison.html">Terminal state identification</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_driver_prediction.html">Early driver perturbation prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_perturbation.html">TFs perturbation prediction with RegVelo</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_driver_prediction_comparison.html">Driver ranking benchmark</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_celloracle.html">CellOracle-based perturbation prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_dynamo.html">Dynamo-based perturbation analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_regvelo_perturbation.html">Using RegVelo to predict perturbation</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_perturbation_benchmark.html">Perturbation prediction benchmark</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Perturb-seq data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_elf1_motif_analysis.html">elf1 perturbation simulation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Smart-seq3 and perturb-seq combined analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/1_Perturb_seq_scanpy.html">Reannotation of pigment cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/2_perturb_seq_ts.html">Evaluate the cell fate perturbation significance on perturb-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/3_regvelo_perturbation.html">Compute density depletion likelihood with RegVelo</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/3_dynamo.html">Dynamo-based perturbation analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/3_celloracle.html">CellOracle-based perturbation prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zebrafish_perturb/4_benchmark.html">Benchmark perturbation prediction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Toy GRN data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/1_simulation.html">Simulation of toy GRN and cellular dynamics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_regvelo.html">RegVelo benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_scvelo.html">scVelo benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_velovi.html">veloVI benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_dpt.html">DPT benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_celloracle.html">CellOracle benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_grnboost2.html">GRNBoost2 benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/2_correlation.html">Correlation benchmark on toy GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/3_comparison.html">Performance comparison of inference on toy GRN data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../toy_grn/3_plot_toy_simulation_benchmark.html">Plot the results of Toy simulate results</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dyngen-simulated data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../dyngen/1_simulation.html">Simulate synthetic datasets with dyngen</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/2_data_conversion.html">Dyngen data preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_regvelo.html">RegVelo benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_scvelo.html">scVelo benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_velovi.html">veloVI benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_unitvelo.html">UniTVelo benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_velovae_fullvb.html">VeloVAE (fullvb) benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_velovae.html">VeloVAE benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_cell2fate.html">cell2fate benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_celloracle.html">CellOracle benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_grnboost2.html">GRNBoost2 benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_splicejac.html">spliceJAC benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/3_correlation.html">Correlation benchmark on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/4_comparison.html">Performance comparison of inference on dyngen data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/5_velovi_benchmark.html">Using veloVI as baseline for RegVelo identifiability test</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/5_regvelo_benchmark.html">Prior GRN sensitivity evaluation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dyngen/5_robustness_evaluation.html">Plot downstream metrics</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/theislab/regvelo_reproducibility" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/zebrafish/8_perturb_seq_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Perturb-seq data analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#library-imports">Library imports</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-setting">General setting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#constants">Constants</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-loading">Data loading</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-via-umap">Visualize via UMAP</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-nr2f2-tfec-mitfa-and-elf1-perturbation-effect">Visualize nr2f2, tfec, mitfa and elf1 perturbation effect</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-tfec-perturbation-panel">Visualize differential expressed genes in tfec perturbation panel</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#test-significance">test significance</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-elf1-perturbation-panel">Visualize differential expressed genes in elf1 perturbation panel</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">test significance</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-pigment-cell-lineage-of-elf1-perturbation-panel">Visualize differential expressed genes in pigment cell lineage of elf1 perturbation panel</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-elf1-gene-expression-change-when-knockout-tfec-and-fli1a">Visualize elf1 gene expression change when knockout tfec and fli1a</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#knockout-tfec">knockout tfec</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">test significance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#knockout-fli1a">knockout fli1a</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">test significance</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="perturb-seq-data-analysis">
<h1>Perturb-seq data analysis<a class="headerlink" href="#perturb-seq-data-analysis" title="Link to this heading">#</a></h1>
<p>Notebook for analyzing zebrafish Perturb-seq data.</p>
<section id="library-imports">
<h2>Library imports<a class="headerlink" href="#library-imports" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">mplscience</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scvelo</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">scv</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">rgv_tools</span><span class="w"> </span><span class="kn">import</span> <span class="n">DATA_DIR</span><span class="p">,</span> <span class="n">FIG_DIR</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.
  warnings.warn(msg, FutureWarning)
/home/icb/weixu.wang/miniconda3/envs/regvelo_test/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.
  warnings.warn(msg, FutureWarning)
</pre></div>
</div>
</div>
</div>
</section>
<section id="general-setting">
<h2>General setting<a class="headerlink" href="#general-setting" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;svg.fonttype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">reset_defaults</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">reset_orig</span><span class="p">()</span>
<span class="n">scv</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="s2">&quot;scvelo&quot;</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="constants">
<h2>Constants<a class="headerlink" href="#constants" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DATASET</span> <span class="o">=</span> <span class="s2">&quot;zebrafish&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SAVE_FIGURES</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
    <span class="p">(</span><span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-loading">
<h2>Data loading<a class="headerlink" href="#data-loading" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;raw&quot;</span> <span class="o">/</span> <span class="s2">&quot;seu_NC_clustered_normalized.h5ad&quot;</span><span class="p">)</span>
<span class="n">ss3</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;processed&quot;</span> <span class="o">/</span> <span class="s2">&quot;adata_preprocessed.h5ad&quot;</span><span class="p">)</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;raw&quot;</span> <span class="o">/</span> <span class="s2">&quot;metadata.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">metadata</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualize-via-umap">
<h2>Visualize via UMAP<a class="headerlink" href="#visualize-via-umap" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Using MELD to calculate sample-associated density estimates and relative likelihood</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
<span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;genotype_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span>
<span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;genotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span>
<span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;replicate&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
<span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>orig.ident</th>
      <th>nCount_RNA</th>
      <th>nFeature_RNA</th>
      <th>sample</th>
      <th>percent.mt</th>
      <th>RNA_snn_res.1</th>
      <th>seurat_clusters</th>
      <th>sgRNA_group</th>
      <th>DoubletFinder</th>
      <th>proj</th>
      <th>RNA_snn_res.0.8</th>
      <th>cell_anno</th>
      <th>latent_time_imputed</th>
      <th>latent_time_imputed_scvi</th>
      <th>term_state_imputed</th>
      <th>cell_anno_old</th>
      <th>genotype_name</th>
      <th>genotype</th>
      <th>replicate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>p1_AAACGGGAGCTAGTTC-1</th>
      <td>SeuratProject</td>
      <td>9374.373090</td>
      <td>3125</td>
      <td>p1</td>
      <td>0.601578</td>
      <td>14</td>
      <td>14</td>
      <td>fli1a_erf_erfl3</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>6</td>
      <td>mNC_head_mesenchymal</td>
      <td>0.599100</td>
      <td>0.567514</td>
      <td>NaN</td>
      <td>mNC_head_mesenchymal</td>
      <td>fli1a_erf_erfl3</td>
      <td>fli1a_erf_erfl3</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p1_AAACGGGCAAAGAATC-1</th>
      <td>SeuratProject</td>
      <td>16617.288939</td>
      <td>3908</td>
      <td>p1</td>
      <td>0.752648</td>
      <td>2</td>
      <td>2</td>
      <td>fli1a_erf_erfl3</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>9</td>
      <td>Pigment_gch2_high</td>
      <td>0.761031</td>
      <td>0.783371</td>
      <td>NaN</td>
      <td>Pigment</td>
      <td>fli1a_erf_erfl3</td>
      <td>fli1a_erf_erfl3</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p1_AAACGGGCAATGTAAG-1</th>
      <td>SeuratProject</td>
      <td>10564.695306</td>
      <td>3262</td>
      <td>p1</td>
      <td>0.888390</td>
      <td>17</td>
      <td>17</td>
      <td>mitfa_tfec_inhouse</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>4</td>
      <td>mNC_hox34</td>
      <td>0.364584</td>
      <td>0.289242</td>
      <td>NaN</td>
      <td>unclassified2</td>
      <td>mitfa_tfec_inhouse</td>
      <td>mitfa_tfec_inhouse</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p1_AAACGGGCAGCAGTTT-1</th>
      <td>SeuratProject</td>
      <td>17754.127675</td>
      <td>3934</td>
      <td>p1</td>
      <td>0.579913</td>
      <td>19</td>
      <td>19</td>
      <td>fli1a_erf_erfl3</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>7</td>
      <td>mNC_arch2</td>
      <td>0.236565</td>
      <td>0.428227</td>
      <td>NaN</td>
      <td>mNC_arch2</td>
      <td>fli1a_erf_erfl3</td>
      <td>fli1a_erf_erfl3</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p1_AAACGGGGTAGCGTAG-1</th>
      <td>SeuratProject</td>
      <td>15028.838200</td>
      <td>3638</td>
      <td>p1</td>
      <td>0.467387</td>
      <td>6</td>
      <td>6</td>
      <td>fli1a_erf_erfl3</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>5</td>
      <td>unclassified1</td>
      <td>0.268428</td>
      <td>0.410988</td>
      <td>NaN</td>
      <td>unclassified1</td>
      <td>fli1a_erf_erfl3</td>
      <td>fli1a_erf_erfl3</td>
      <td>A</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>p11_TTTGCGCAGAGTAATC-1</th>
      <td>SeuratProject</td>
      <td>6202.170676</td>
      <td>2499</td>
      <td>p11</td>
      <td>0.776390</td>
      <td>8</td>
      <td>8</td>
      <td>ebf3a</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>0</td>
      <td>mNC_vagal</td>
      <td>0.387833</td>
      <td>0.428689</td>
      <td>NaN</td>
      <td>Pigment_sox6_high_vagel</td>
      <td>ebf3a</td>
      <td>ebf3a</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p11_TTTGGTTTCAACTCTT-1</th>
      <td>SeuratProject</td>
      <td>3644.718485</td>
      <td>1857</td>
      <td>p11</td>
      <td>1.794744</td>
      <td>5</td>
      <td>5</td>
      <td>elf1</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>6</td>
      <td>mNC_head_mesenchymal</td>
      <td>0.538755</td>
      <td>0.549123</td>
      <td>NaN</td>
      <td>mNC_head_mesenchymal</td>
      <td>elf1</td>
      <td>elf1</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p11_TTTGGTTTCCTATTCA-1</th>
      <td>SeuratProject</td>
      <td>8422.826902</td>
      <td>3344</td>
      <td>p11</td>
      <td>0.510019</td>
      <td>25</td>
      <td>25</td>
      <td>nr2f2</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>5</td>
      <td>mNC_arch1</td>
      <td>0.412674</td>
      <td>0.363753</td>
      <td>NaN</td>
      <td>mNC_arch1</td>
      <td>nr2f2</td>
      <td>nr2f2</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p11_TTTGGTTTCTAACTCT-1</th>
      <td>SeuratProject</td>
      <td>6612.583387</td>
      <td>2661</td>
      <td>p11</td>
      <td>0.449287</td>
      <td>6</td>
      <td>6</td>
      <td>ebf3a</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>15</td>
      <td>unclassified1</td>
      <td>0.259958</td>
      <td>0.374038</td>
      <td>NaN</td>
      <td>unclassified1</td>
      <td>ebf3a</td>
      <td>ebf3a</td>
      <td>A</td>
    </tr>
    <tr>
      <th>p11_TTTGTCAAGGACAGCT-1</th>
      <td>SeuratProject</td>
      <td>11372.933636</td>
      <td>3694</td>
      <td>p11</td>
      <td>0.594269</td>
      <td>14</td>
      <td>14</td>
      <td>multiplet</td>
      <td>Singlet</td>
      <td>perturbseq</td>
      <td>6</td>
      <td>mNC_head_mesenchymal</td>
      <td>0.489412</td>
      <td>0.520055</td>
      <td>NaN</td>
      <td>mNC_head_mesenchymal</td>
      <td>multiplet</td>
      <td>multiplet</td>
      <td>A</td>
    </tr>
  </tbody>
</table>
<p>12393 rows × 19 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Update pigment cell annotation</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Pigment_gch2_high&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Pigment&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="s2">&quot;mitfa_tfec_inhouse&quot;</span><span class="p">,</span>
            <span class="s2">&quot;negative&quot;</span><span class="p">,</span>
            <span class="s2">&quot;control&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mitfa&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tfec&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tfec_mitfa_bhlhe40&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mitfa_tfec&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mitfa_tfec_tfeb&quot;</span><span class="p">,</span>
            <span class="s2">&quot;elf1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nr2f2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nr2f5&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">highlight_cell_types</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;mNC_hox34&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mNC_head_mesenchymal&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mNC_arch2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Pigment&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Other&quot;</span><span class="p">,</span>
<span class="p">]</span>  <span class="c1"># replace with your cell types of interest</span>

<span class="c1"># Create a new column for colors</span>
<span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;highlight&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">palette</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">ss3</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span> <span class="n">ss3</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;cell_type_colors&quot;</span><span class="p">]))</span>
<span class="c1"># Set cell types to grey if they are not in the highlight list</span>
<span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;highlight&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;highlight&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">highlight_cell_types</span> <span class="k">else</span> <span class="s2">&quot;Other&quot;</span><span class="p">)</span>

<span class="c1"># Create a color map where &#39;Other&#39; is grey</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;grey&quot;</span> <span class="k">if</span> <span class="n">ct</span> <span class="o">==</span> <span class="s2">&quot;Other&quot;</span> <span class="k">else</span> <span class="n">palette</span><span class="p">[</span><span class="n">ct</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ct</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">highlight_cell_types</span><span class="p">)]</span>
<span class="n">color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">ct</span><span class="p">:</span> <span class="n">color</span> <span class="k">for</span> <span class="n">ct</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">highlight_cell_types</span><span class="p">,</span> <span class="n">colors</span><span class="p">)}</span>
<span class="n">color_map</span><span class="p">[</span><span class="s2">&quot;Other&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;grey&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
        <span class="n">perturbseq</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;highlight&quot;</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;phate&quot;</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">color_map</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;INTRO_figure_perturb.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/dc6417570024e253749e61079c279c2f2eba842b0db4e549f839e2e49534e908.png"><img alt="../_images/dc6417570024e253749e61079c279c2f2eba842b0db4e549f839e2e49534e908.png" src="../_images/dc6417570024e253749e61079c279c2f2eba842b0db4e549f839e2e49534e908.png" style="width: 452px; height: 341px;" />
</a>
</div>
</div>
</section>
<section id="visualize-nr2f2-tfec-mitfa-and-elf1-perturbation-effect">
<h2>Visualize nr2f2, tfec, mitfa and elf1 perturbation effect<a class="headerlink" href="#visualize-nr2f2-tfec-mitfa-and-elf1-perturbation-effect" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sgRNA_groups</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;nr2f2&quot;</span><span class="p">,</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;mitfa&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the perturbation on terminal states</span>
<span class="n">cell_types</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;mNC_hox34&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mNC_head_mesenchymal&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mNC_arch2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Pigment&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">sgRNA_groups</span><span class="p">:</span>
    <span class="n">sample_likelihoods</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;raw&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;likelihood_</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">condition</span> <span class="o">=</span> <span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;control&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">group</span><span class="p">)</span>
    <span class="n">experimental_samples</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">A&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">B&quot;</span><span class="p">]</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sample_likelihoods</span><span class="p">[</span><span class="n">experimental_samples</span><span class="p">])</span>
    <span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># Calculate the percentiles values</span>

    <span class="n">down</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">up</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.8</span><span class="p">)</span>

    <span class="c1"># Apply the condition to set values outside the 5th and 95th percentiles to NaN</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">up</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">down</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">group</span><span class="p">])</span>

    <span class="n">vec</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_types</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">])</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vec</span><span class="p">)</span>

    <span class="n">adata_sub</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;quantile_value&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1">## Plot the figure</span>
    <span class="n">overall_score</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span> <span class="o">*</span> <span class="n">adata_sub</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])})</span>
    <span class="n">overall_score</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">overall_score</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="s2">&quot;score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">])</span>

    <span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">overall_score</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">])</span>
    <span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;highlight&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">]))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="n">highlighted_points</span> <span class="o">=</span> <span class="n">adata_sub</span><span class="p">[</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;highlight&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;True&quot;</span><span class="p">]</span>
    <span class="n">x_highlight</span> <span class="o">=</span> <span class="n">highlighted_points</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_phate&quot;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">y_highlight</span> <span class="o">=</span> <span class="n">highlighted_points</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_phate&quot;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">color_values</span> <span class="o">=</span> <span class="n">highlighted_points</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">]</span>  <span class="c1"># Adjust to your continuous variable</span>

    <span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_sub</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;phate&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_score&quot;</span><span class="p">,</span> <span class="n">vcenter</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">x_highlight</span><span class="p">,</span>
            <span class="n">y_highlight</span><span class="p">,</span>
            <span class="n">c</span><span class="o">=</span><span class="n">color_values</span><span class="p">,</span>
            <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span>  <span class="c1"># Choose a gradient color map, e.g., &#39;viridis&#39;, &#39;plasma&#39;, &#39;inferno&#39;, etc.</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>  <span class="c1"># Larger size for highlighted points</span>
            <span class="c1"># edgecolor=&#39;black&#39;,  # Optional: border for emphasis</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
                <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">_perturbation.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span>
            <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/37ea0d92fc0f0cb1c6e22b224aba02a24eb69edda3d9881c98e8758c23c64b1a.png"><img alt="../_images/37ea0d92fc0f0cb1c6e22b224aba02a24eb69edda3d9881c98e8758c23c64b1a.png" src="../_images/37ea0d92fc0f0cb1c6e22b224aba02a24eb69edda3d9881c98e8758c23c64b1a.png" style="width: 436px; height: 303px;" />
</a>
<a class="reference internal image-reference" href="../_images/f0bbaf1c26586205607d3b329091e93a800ea2047a134e66e75b6e2b6c23ec38.png"><img alt="../_images/f0bbaf1c26586205607d3b329091e93a800ea2047a134e66e75b6e2b6c23ec38.png" src="../_images/f0bbaf1c26586205607d3b329091e93a800ea2047a134e66e75b6e2b6c23ec38.png" style="width: 436px; height: 303px;" />
</a>
<a class="reference internal image-reference" href="../_images/7e7a6d8b2c593cb02f2635c76454d0a45920093e16ef6f6540f6b572ac86d855.png"><img alt="../_images/7e7a6d8b2c593cb02f2635c76454d0a45920093e16ef6f6540f6b572ac86d855.png" src="../_images/7e7a6d8b2c593cb02f2635c76454d0a45920093e16ef6f6540f6b572ac86d855.png" style="width: 436px; height: 303px;" />
</a>
<a class="reference internal image-reference" href="../_images/86f14e8c2aba238ed502b513ecbee4362b68b355f29ff51d22d4132421b6e154.png"><img alt="../_images/86f14e8c2aba238ed502b513ecbee4362b68b355f29ff51d22d4132421b6e154.png" src="../_images/86f14e8c2aba238ed502b513ecbee4362b68b355f29ff51d22d4132421b6e154.png" style="width: 436px; height: 303px;" />
</a>
</div>
</div>
</section>
<section id="visualize-differential-expressed-genes-in-tfec-perturbation-panel">
<h2>Visualize differential expressed genes in tfec perturbation panel<a class="headerlink" href="#visualize-differential-expressed-genes-in-tfec-perturbation-panel" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">perturbseq</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span>
    <span class="o">~</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;unclassified2&quot;</span><span class="p">,</span> <span class="s2">&quot;unclassified1&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant_hox23&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant&quot;</span><span class="p">]),</span>
<span class="p">]</span>
<span class="n">seu_small</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="s2">&quot;tfec&quot;</span><span class="p">]),]</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">seu_small</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;cdkn1ca&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;cdkn1ca.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/42924cdaadb34da9e0726c06714f4ab7d752a6d4e1457fe7183ba95d079746c6.png"><img alt="../_images/42924cdaadb34da9e0726c06714f4ab7d752a6d4e1457fe7183ba95d079746c6.png" src="../_images/42924cdaadb34da9e0726c06714f4ab7d752a6d4e1457fe7183ba95d079746c6.png" style="width: 239px; height: 243px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;atp6ap2&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;atp6ap2.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/831b0d57cd858d7d4803577b40cd77aa4f0c7fc2bd35a8b703b50ae33e265f43.png"><img alt="../_images/831b0d57cd858d7d4803577b40cd77aa4f0c7fc2bd35a8b703b50ae33e265f43.png" src="../_images/831b0d57cd858d7d4803577b40cd77aa4f0c7fc2bd35a8b703b50ae33e265f43.png" style="width: 239px; height: 243px;" />
</a>
</div>
</div>
<section id="test-significance">
<h3>test significance<a class="headerlink" href="#test-significance" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;cdkn1ca&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;cdkn1ca&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;less&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=-2.0416193271649687, pvalue=0.021191380003823368, df=219.9875671451483)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;atp6ap2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;atp6ap2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;less&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=-5.547499181077916, pvalue=3.4438545499844835e-08, df=270.9981680245929)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="visualize-differential-expressed-genes-in-elf1-perturbation-panel">
<h2>Visualize differential expressed genes in elf1 perturbation panel<a class="headerlink" href="#visualize-differential-expressed-genes-in-elf1-perturbation-panel" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">perturbseq</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span>
    <span class="o">~</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;unclassified2&quot;</span><span class="p">,</span> <span class="s2">&quot;unclassified1&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant_hox23&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant&quot;</span><span class="p">]),</span>
<span class="p">]</span>
<span class="n">seu_small</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]),]</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">seu_small</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;mitfa&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;mitfa_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/7e8ac568d4a38804e162cd9fc957197c7e9080bb9d0c60c02ddcfd91ba6a0f05.png"><img alt="../_images/7e8ac568d4a38804e162cd9fc957197c7e9080bb9d0c60c02ddcfd91ba6a0f05.png" src="../_images/7e8ac568d4a38804e162cd9fc957197c7e9080bb9d0c60c02ddcfd91ba6a0f05.png" style="width: 239px; height: 243px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;sox10&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;sox10_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/b56a6ec72f31c840f82c4767e98d34530f78521ad83293e3c6e7c1fce2c55729.png"><img alt="../_images/b56a6ec72f31c840f82c4767e98d34530f78521ad83293e3c6e7c1fce2c55729.png" src="../_images/b56a6ec72f31c840f82c4767e98d34530f78521ad83293e3c6e7c1fce2c55729.png" style="width: 239px; height: 243px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;fli1a&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;fli1a_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/49866fc5eed286dd0e2a3e30019fc8592b2846a619bfd0f7d67e0150b093cfdd.png"><img alt="../_images/49866fc5eed286dd0e2a3e30019fc8592b2846a619bfd0f7d67e0150b093cfdd.png" src="../_images/49866fc5eed286dd0e2a3e30019fc8592b2846a619bfd0f7d67e0150b093cfdd.png" style="width: 251px; height: 238px;" />
</a>
</div>
</div>
<section id="id1">
<h3>test significance<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;mitfa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;mitfa&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;less&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=-3.429003795549063, pvalue=0.0003452279743702678, df=300.522449116822)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;sox10&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;sox10&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;less&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=-4.108995730134873, pvalue=2.6789127768845648e-05, df=255.2575590451451)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;fli1a&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;elf1&quot;</span><span class="p">,</span> <span class="s2">&quot;fli1a&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;greater&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=1.3860128076275413, pvalue=0.0835617324921002, df=223.63128948496157)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="visualize-differential-expressed-genes-in-pigment-cell-lineage-of-elf1-perturbation-panel">
<h2>Visualize differential expressed genes in pigment cell lineage of elf1 perturbation panel<a class="headerlink" href="#visualize-differential-expressed-genes-in-pigment-cell-lineage-of-elf1-perturbation-panel" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">perturbseq</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;Pigment&quot;</span><span class="p">]),]</span>
<span class="n">seu_small</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]),]</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">seu_small</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;pmp22a&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;pmp22a_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/06d5ae9098da2c819311f12b6ff55a000854cf41b3b238c2e5901f3edaba2735.png"><img alt="../_images/06d5ae9098da2c819311f12b6ff55a000854cf41b3b238c2e5901f3edaba2735.png" src="../_images/06d5ae9098da2c819311f12b6ff55a000854cf41b3b238c2e5901f3edaba2735.png" style="width: 239px; height: 238px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;cited1&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;cited1_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/cd91b8a2f1830e983cd043f4dd52d70237ba1fec47086deec16929a8a32e3a28.png"><img alt="../_images/cd91b8a2f1830e983cd043f4dd52d70237ba1fec47086deec16929a8a32e3a28.png" src="../_images/cd91b8a2f1830e983cd043f4dd52d70237ba1fec47086deec16929a8a32e3a28.png" style="width: 251px; height: 238px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;hmgn2&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;hmgn2_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/ea57db1348ec037f5f956d0f566d3845abb19608b2d01d3ad156f7b30740ef2b.png"><img alt="../_images/ea57db1348ec037f5f956d0f566d3845abb19608b2d01d3ad156f7b30740ef2b.png" src="../_images/ea57db1348ec037f5f956d0f566d3845abb19608b2d01d3ad156f7b30740ef2b.png" style="width: 239px; height: 238px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;arl6ip1&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;arl6ip1_expression_elf1_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/7be389076d324c2b0744099d184b9f1071a1e9a9300928e88a38eca9e7a16e83.png"><img alt="../_images/7be389076d324c2b0744099d184b9f1071a1e9a9300928e88a38eca9e7a16e83.png" src="../_images/7be389076d324c2b0744099d184b9f1071a1e9a9300928e88a38eca9e7a16e83.png" style="width: 251px; height: 238px;" />
</a>
</div>
</div>
</section>
<section id="visualize-elf1-gene-expression-change-when-knockout-tfec-and-fli1a">
<h2>Visualize elf1 gene expression change when knockout tfec and fli1a<a class="headerlink" href="#visualize-elf1-gene-expression-change-when-knockout-tfec-and-fli1a" title="Link to this heading">#</a></h2>
<section id="knockout-tfec">
<h3>knockout tfec<a class="headerlink" href="#knockout-tfec" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">perturbseq</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span>
    <span class="o">~</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;unclassified2&quot;</span><span class="p">,</span> <span class="s2">&quot;unclassified1&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant_hox23&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant&quot;</span><span class="p">]),</span>
<span class="p">]</span>
<span class="n">seu_small</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="s2">&quot;tfec&quot;</span><span class="p">]),]</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">seu_small</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;elf1&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;elf1_expression_tfec_knockout.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/cd11c12ae7d090c54cc0ede5baa71fdff9bd8b005ad23619dbc3148f0e848edd.png"><img alt="../_images/cd11c12ae7d090c54cc0ede5baa71fdff9bd8b005ad23619dbc3148f0e848edd.png" src="../_images/cd11c12ae7d090c54cc0ede5baa71fdff9bd8b005ad23619dbc3148f0e848edd.png" style="width: 251px; height: 238px;" />
</a>
</div>
</div>
</section>
<section id="id2">
<h3>test significance<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;tfec&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;less&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=-5.366020270431321, pvalue=8.610173685738961e-08, df=272.49193453818)
</pre></div>
</div>
</div>
</div>
</section>
<section id="knockout-fli1a">
<h3>knockout fli1a<a class="headerlink" href="#knockout-fli1a" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbseq</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">perturbseq</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span>
    <span class="o">~</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_anno_new&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;unclassified2&quot;</span><span class="p">,</span> <span class="s2">&quot;unclassified1&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant_hox23&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant&quot;</span><span class="p">]),</span>
<span class="p">]</span>
<span class="n">seu_small</span> <span class="o">=</span> <span class="n">perturbseq</span><span class="p">[</span><span class="n">perturbseq</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="s2">&quot;fli1a&quot;</span><span class="p">]),]</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">seu_small</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">mplscience</span><span class="o">.</span><span class="n">style_context</span><span class="p">():</span>  <span class="c1"># Use the mplscience style context</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Create the scatter plot</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">seu_small</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;elf1&quot;</span><span class="p">],</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;sgRNA_group&quot;</span><span class="p">,</span> <span class="n">stripplot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">SAVE_FIGURES</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
            <span class="n">FIG_DIR</span> <span class="o">/</span> <span class="n">DATASET</span> <span class="o">/</span> <span class="s2">&quot;elf1_expression_fli1a_knockout.svg&quot;</span><span class="p">,</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span>
            <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Display the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/9b9c3ae546e3cccba0e78a64e9a6721b5daaaef56b704e3e6d973d993762597d.png"><img alt="../_images/9b9c3ae546e3cccba0e78a64e9a6721b5daaaef56b704e3e6d973d993762597d.png" src="../_images/9b9c3ae546e3cccba0e78a64e9a6721b5daaaef56b704e3e6d973d993762597d.png" style="width: 251px; height: 238px;" />
</a>
</div>
</div>
</section>
</section>
<section id="id3">
<h2>test significance<a class="headerlink" href="#id3" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;fli1a&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">seu_small</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;fli1a&quot;</span><span class="p">,</span> <span class="s2">&quot;elf1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;greater&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TtestResult(statistic=3.059975496268048, pvalue=0.0011362106276185085, df=1002.4934192631699)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "conda-env-regvelo_test-py"
        },
        kernelOptions: {
            name: "conda-env-regvelo_test-py",
            path: "./zebrafish"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'conda-env-regvelo_test-py'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="7_perturbation_benchmark.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Perturbation prediction benchmark</p>
      </div>
    </a>
    <a class="right-next"
       href="9_elf1_motif_analysis.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">elf1 perturbation simulation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#library-imports">Library imports</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-setting">General setting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#constants">Constants</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-loading">Data loading</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-via-umap">Visualize via UMAP</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-nr2f2-tfec-mitfa-and-elf1-perturbation-effect">Visualize nr2f2, tfec, mitfa and elf1 perturbation effect</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-tfec-perturbation-panel">Visualize differential expressed genes in tfec perturbation panel</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#test-significance">test significance</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-elf1-perturbation-panel">Visualize differential expressed genes in elf1 perturbation panel</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">test significance</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-differential-expressed-genes-in-pigment-cell-lineage-of-elf1-perturbation-panel">Visualize differential expressed genes in pigment cell lineage of elf1 perturbation panel</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-elf1-gene-expression-change-when-knockout-tfec-and-fli1a">Visualize elf1 gene expression change when knockout tfec and fli1a</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#knockout-tfec">knockout tfec</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">test significance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#knockout-fli1a">knockout fli1a</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">test significance</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Weixu Wang, Philipp Weiler
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>